pkg/core/transaction/witness_condition_test.go

```diff
@@ -0,0 +1,318 @@
+	require.NoError(t, err)
+	pkNeutral, err := keys.NewPrivateKey()
+	require.NoError(t, err)
+	entrySC := util.Uint160{1, 2, 3}
+	currentSC := util.Uint160{4, 5, 6}
+	tmc := &TestMC{
+		calling: entrySC,
+		entry:   entrySC,
+		current: currentSC,
+		goodKey: pkGood.PublicKey(),
+		badKey:  pkBad.PublicKey(),
+	}
+
+	t.Run("boolean", func(t *testing.T) {
+		var b bool
+		var c = (*ConditionBoolean)(&b)
+		res, err := c.Match(tmc)
+		require.NoError(t, err)
+		require.False(t, res)
+		b = true
+		res, err = c.Match(tmc)
+		require.NoError(t, err)
+		require.True(t, res)
+	})
+	t.Run("not", func(t *testing.T) {
+		var b bool
+		var cInner = (*ConditionBoolean)(&b)
+		var cInner2 = (*ConditionGroup)(pkBad.PublicKey())
+		var c = &ConditionNot{cInner}
+		var c2 = &ConditionNot{cInner2}
+
+		res, err := c.Match(tmc)
+		require.NoError(t, err)
+		require.True(t, res)
+		b = true
+		res, err = c.Match(tmc)
+		require.NoError(t, err)
+		require.False(t, res)
+		_, err = c2.Match(tmc)
+		require.Error(t, err)
+	})
+	t.Run("and", func(t *testing.T) {
+		var bFalse, bTrue bool
+		var cInnerFalse = (*ConditionBoolean)(&bFalse)
+		var cInnerTrue = (*ConditionBoolean)(&bTrue)
+		var cInnerBad = (*ConditionGroup)(pkBad.PublicKey())
+		var c = &ConditionAnd{cInnerTrue, cInnerFalse, cInnerFalse}
+		var cBad = &ConditionAnd{cInnerTrue, cInnerBad}
+
+		bTrue = true
+		res, err := c.Match(tmc)
+		require.NoError(t, err)
+		require.False(t, res)
+		bFalse = true
+		res, err = c.Match(tmc)
+		require.NoError(t, err)
+		require.True(t, res)
+
+		_, err = cBad.Match(tmc)
+		require.Error(t, err)
+	})
+	t.Run("or", func(t *testing.T) {
+		var bFalse, bTrue bool
+		var cInnerFalse = (*ConditionBoolean)(&bFalse)
+		var cInnerTrue = (*ConditionBoolean)(&bTrue)
+		var cInnerBad = (*ConditionGroup)(pkBad.PublicKey())
+		var c = &ConditionOr{cInnerTrue, cInnerFalse, cInnerFalse}
+		var cBad = &ConditionOr{cInnerTrue, cInnerBad}
+
+		bTrue = true
+		res, err := c.Match(tmc)
+		require.NoError(t, err)
+		require.True(t, res)
+		bTrue = false
+		res, err = c.Match(tmc)
+		require.NoError(t, err)
+		require.False(t, res)
+
+		_, err = cBad.Match(tmc)
+		require.Error(t, err)
+	})
+	t.Run("script hash", func(t *testing.T) {
+		var cEntry = (*ConditionScriptHash)(&entrySC)
+		var cCurrent = (*ConditionScriptHash)(&currentSC)
+
+		res, err := cEntry.Match(tmc)
+		require.NoError(t, err)
+		require.False(t, res)
+		res, err = cCurrent.Match(tmc)
+		require.NoError(t, err)
+		require.True(t, res)
+	})
+	t.Run("group", func(t *testing.T) {
+		var cBad = (*ConditionGroup)(pkBad.PublicKey())
+		var cGood = (*ConditionGroup)(pkGood.PublicKey())
+		var cNeutral = (*ConditionGroup)(pkNeutral.PublicKey())
+
+		res, err := cGood.Match(tmc)
+		require.NoError(t, err)
+		require.True(t, res)
+
+		res, err = cNeutral.Match(tmc)
+		require.NoError(t, err)
+		require.False(t, res)
+
+		_, err = cBad.Match(tmc)
+		require.Error(t, err)
+	})
+	t.Run("called by entry", func(t *testing.T) {
```

Comment:
We can add the case with `calling == util.Uint160{} && current == entry`, just for completeness.